pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+// CompileRegex provides a convenience wrapper around the LRU. It will either
+// return the cached regex object or compile the pattern into a regex object if
+// the LRU hasn't been initialized (and initializes it) or it doesn't exist in
+// the cache. If it doesn't exist in the cache, it will be added.
+//
+// This function assumes that it's invoked after the LRU has been initialized
+// via InitRegexCompileLRU(), which happens during the DNS proxy
+// initialization, which assumes that the K8s Watcher is not yet synced and no
+// policies are being processed by the system (meaning no policies to parse,
+// i.e. compiling the FQDN rules or handling FQDN selectors). If that doesn't
+// hold true, then we attempt a best-effort to initialize the LRU.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
+		// Try to initialize the LRU here in case the caller never called
+		// InitRegexCompileLRU() explicitly.
+		initOnce.Do(func() {
+			if err := InitRegexCompileLRU(option.Config.FQDNRegexCompileLRUSize); err != nil {
+				log.WithError(err).Warn(
+					"Failed to initialize FQDN regex LRU. Cilium will not cache FQDN regexes. Please report this if you see this message.",
+				)
+			}
+		})
+		// Since this is the case where the LRU is not initialized, we accept
+		// that we won't cache the first result here.
+		return regexp.Compile(p)
+	}
+	RegexCompileLRU.Lock()
+	r, ok := RegexCompileLRU.Get(p)
+	RegexCompileLRU.Unlock()
```

Comment:
Should this take out just a read lock to avoid blocking on read only cache hit happy path?

```suggestion
	RegexCompileLRU.RLock()
	r, ok := RegexCompileLRU.Get(p)
	RegexCompileLRU.RUnlock()
```

Reply:
I think that the concern here is that since the underlying structure is an LRU, presumably the `Get()` operation also modifies the underlying order that is stored, so readlocks would not be safe for concurrent access.

Overall though I presume that this will be primarily used during policy calculation and then the caller stores the underlying regex object for regular usage against the target requests, so it shouldn't be too bad even with a regular mutex.

Reply:
^ Correct. If you look at the `Get()` call, you can see it's potentially modifying the underlying structure to evict the oldest member, etc.

Reply:
Ahh, good call